File: backend/src/versionEngine.js
const fs = require('fs-extra');
const path = require('path');
const { v4: uuidv4 } = require('uuid');
const { storageManager } = require('./storageManager');
const fossilDelta = require('fossil-delta');
const zlib = require('zlib');

const SKIP_GZIP_EXT = new Set([
  '.zip', '.rar', '.7z', '.tar', '.gz',
  '.mp4', '.mkv', '.avi', '.mov',
  '.mp3', '.flac', '.aac',
  '.jpg', '.jpeg', '.png', '.gif', '.webp'
]);

// File extensions to prioritize
const HIGH_PRIORITY_EXT = new Set([
  '.js', '.jsx', '.ts', '.tsx', '.py', '.cpp', '.c', '.h', '.java', '.go', '.rs',
  '.json', '.yaml', '.yml', '.toml', '.env',
  '.pdf', '.docx', '.doc', '.xlsx', '.xls', '.pptx', '.txt', '.md',
  '.jpg', '.jpeg', '.png', '.gif', '.svg', '.webp',
  '.mp4', '.mp3', '.wav', '.zip', '.rar'
]);

// Patterns to ignore
const IGNORE_PATTERNS = [
  '.restorex', 'node_modules', '.git', '$RECYCLE.BIN', 'System Volume Information',
  '.tmp', '.temp', '.log', '.cache', 'Thumbs.db', 'desktop.ini', '.DS_Store',
  '__pycache__', '.pytest_cache', 'dist', '.next', 'build', '.vscode', '.idea'
];

class VersionEngine {
  constructor() {
    this.debounceMap = new Map();
    this.DEBOUNCE_MS = 600;
    this.activeTasks = 0;
    this.pendingRestores = new Map(); // path -> versionId
  }

  isSyncing() {
    return this.activeTasks > 0;
  }

  shouldIgnore(relPath) {
    const lower = relPath.toLowerCase().replace(/\\/g, '/');
    for (const pattern of IGNORE_PATTERNS) {
      if (lower.includes(pattern.toLowerCase())) return true;
    }
    const ext = path.extname(relPath).toLowerCase();
    // Only track files with known extensions or no extension (important docs)
    if (ext && !HIGH_PRIORITY_EXT.has(ext)) {
      // Allow tracking but don't prioritize
    }
    return false;
  }

  handleEvent(action, relPath, watchPath) {
    console.log(`[VersionEngine] Event: ${action} on ${relPath}`);
    if (this.shouldIgnore(relPath)) {
        console.log(`[VersionEngine] Ignored: ${relPath}`);
        return;
    }
    if (!relPath || relPath.trim() === '') return;

    const key = `${action}:${relPath}`;
    if (this.debounceMap.has(key)) {
      clearTimeout(this.debounceMap.get(key));
    } else {
      this.activeTasks++;
    }

    const timer = setTimeout(() => {
      this.debounceMap.delete(key);
      this._process(action, relPath, watchPath)
        .catch(err => {
          console.error('[VersionEngine] error:', err.message);
        })
        .finally(() => {
          this.activeTasks--;
        });
    }, this.DEBOUNCE_MS);

    this.debounceMap.set(key, timer);
  }

  async _process(action, relPath, watchPath) {
    const fullPath = path.join(watchPath, relPath);
    const storageDir = path.join(watchPath, '.restorex', 'versions', path.dirname(relPath));
    const timestamp = new Date().toISOString();
    const safeTs = timestamp.replace(/[:.]/g, '-');
    const basename = path.basename(relPath);
    const storageName = `${safeTs}__${basename}`;
    const storagePath = path.join(storageDir, storageName);

    const versionId = uuidv4();

    if (action === 'created' || action === 'modified') {
      try {
        const stat = await fs.stat(fullPath);
        if (!stat.isFile()) return;

        await fs.ensureDir(storageDir);

        let encoding = 'full';
        let parentVersionId = null;
        const fileRecord = storageManager.getFileByPath(relPath);

        const ext = path.extname(relPath).toLowerCase();
        const canGzip = stat.size <= 50 * 1024 * 1024 && !SKIP_GZIP_EXT.has(ext);

        const saveFullOrGzip = async (src, dst) => {
          if (canGzip) {
            const buffer = await fs.readFile(src);
            const gzipped = zlib.gzipSync(buffer);
            if (gzipped.length < buffer.length) {
              await fs.writeFile(dst, gzipped);
              return 'gzip';
            }
          }
          await fs.copy(src, dst);
          return 'full';
        };

        // Try to do delta compression if we have a previous version and size is <= 50MB
        if (fileRecord && fileRecord.versions.length > 0 && stat.size <= 50 * 1024 * 1024) {
          let lastActive = null;
          for (let i = fileRecord.versions.length - 1; i >= 0; i--) {
            if (fileRecord.versions[i].status !== 'deleted' && fileRecord.versions[i].storagePath) {
              lastActive = fileRecord.versions[i];
              break;
            }
          }

          if (lastActive) {
            try {
              const prevBuffer = await this._reconstructVersion(fileRecord, lastActive.versionId);
              const newBuffer = await fs.readFile(fullPath);
              const deltaArray = fossilDelta.createDelta(prevBuffer, newBuffer);
              
              // Only use delta if it's actually smaller than the full file
              if (deltaArray.length < newBuffer.length) {
                await fs.writeFile(storagePath, Buffer.from(deltaArray));
                encoding = 'delta';
                parentVersionId = lastActive.versionId;
                console.log(`[VersionEngine] Delta size: ${deltaArray.length} bytes (Full: ${newBuffer.length} bytes)`);
              } else {
                encoding = await saveFullOrGzip(fullPath, storagePath);
              }
            } catch (err) {
              console.warn('[VersionEngine] Delta creation failed, falling back to full copy:', err.message);
              encoding = await saveFullOrGzip(fullPath, storagePath);
            }
          } else {
            encoding = await saveFullOrGzip(fullPath, storagePath);
          }
        } else {
          encoding = await saveFullOrGzip(fullPath, storagePath);
        }

        const restoredFrom = this.pendingRestores.get(relPath);
        const finalStatus = restoredFrom ? 'restored' : action;
        this.pendingRestores.delete(relPath);

        await storageManager.addVersion(relPath, {
          versionId,
          timestamp,
          size: stat.size,
          status: finalStatus,
          storagePath,
          restoredFrom,
          encoding,
          parentVersionId
        });

        if (restoredFrom) {
          await storageManager.upsertFile(relPath, { lastRestoredVersionId: restoredFrom });
        }

        console.log(`[+] Versioned ${finalStatus} (${encoding}): ${relPath}`);
      } catch (err) {
        console.warn(`[VersionEngine] Could not version ${relPath}:`, err.message);
      }
    } else if (action === 'deleted') {
      // Mark as deleted — last known copy already saved
      const file = storageManager.getFileByPath(relPath);
      if (file) {
        await storageManager.addVersion(relPath, {
          versionId: uuidv4(),
          timestamp,
          size: file.size || (file.versions.length ? file.versions[file.versions.length - 1].size : 0),
          status: 'deleted',
          storagePath: null,
        });
        console.log(`[-] Marked deleted: ${relPath}`);
      }
    } else if (action === 'renamed_new') {
      // Treat renamed-to as a new file creation
      try {
        const stat = await fs.stat(fullPath);
        if (!stat.isFile()) return;
        const ext = path.extname(relPath).toLowerCase();
        let encoding = 'full';
        if (stat.size <= 50 * 1024 * 1024 && !SKIP_GZIP_EXT.has(ext)) {
          const buffer = await fs.readFile(fullPath);
          const gzipped = zlib.gzipSync(buffer);
          if (gzipped.length < buffer.length) {
            await fs.writeFile(storagePath, gzipped);
            encoding = 'gzip';
          } else {
            await fs.copy(fullPath, storagePath);
          }
        } else {
          await fs.copy(fullPath, storagePath);
        }
        await storageManager.addVersion(relPath, {
          versionId, timestamp, size: stat.size, status: 'created', storagePath, encoding
        });
      } catch (e) {}
    }
  }

  async _reconstructVersion(file, versionId) {
    const versionIdx = file.versions.findIndex(v => v.versionId === versionId);
    if (versionIdx === -1) throw new Error('Version not found');
    
    const version = file.versions[versionIdx];
    if (version.status === 'deleted' || !version.storagePath) throw new Error('Cannot reconstruct deleted version');

    if (version.encoding !== 'delta') {
      if (version.encoding === 'gzip') {
        return zlib.gunzipSync(await fs.readFile(version.storagePath));
      }
      return await fs.readFile(version.storagePath);
    }

    const chain = [];
    let currentIdx = versionIdx;
    while (currentIdx >= 0) {
      const v = file.versions[currentIdx];
      chain.unshift(v);
      if (v.encoding !== 'delta') break;
      currentIdx = file.versions.findIndex(parent => parent.versionId === v.parentVersionId);
      if (currentIdx === -1) throw new Error('Delta chain broken');
    }

    let buffer;
    if (chain[0].encoding === 'gzip') {
      const gzipped = await fs.readFile(chain[0].storagePath);
      buffer = zlib.gunzipSync(gzipped);
    } else {
      buffer = await fs.readFile(chain[0].storagePath);
    }
    for (let i = 1; i < chain.length; i++) {
      const deltaBuffer = await fs.readFile(chain[i].storagePath);
      try {
        const resultArray = fossilDelta.applyDelta(buffer, deltaBuffer);
        buffer = Buffer.from(resultArray);
      } catch (err) {
        throw new Error(`Failed to apply delta at version ${chain[i].versionId}: ${err.message}`);
      }
    }
    return buffer;
  }

  async createBaseline(relPath, watchPath) {
    if (this.shouldIgnore(relPath)) return;
    const fullPath = path.join(watchPath, relPath);
    try {
      const stat = await fs.stat(fullPath);
      if (!stat.isFile()) return;
      if (stat.size > 10 * 1024 * 1024) return; // Skip files > 10MB for baseline

      const storageDir = path.join(watchPath, '.restorex', 'versions', path.dirname(relPath));
      const timestamp = new Date().toISOString();
      const safeTs = timestamp.replace(/[:.]/g, '-');
      const storageName = `${safeTs}__${path.basename(relPath)}`;
      const storagePath = path.join(storageDir, storageName);

      await fs.ensureDir(storageDir);
      const ext = path.extname(relPath).toLowerCase();
      let encoding = 'full';
      if (stat.size <= 50 * 1024 * 1024 && !SKIP_GZIP_EXT.has(ext)) {
        const buffer = await fs.readFile(fullPath);
        const gzipped = zlib.gzipSync(buffer);
        if (gzipped.length < buffer.length) {
          await fs.writeFile(storagePath, gzipped);
          encoding = 'gzip';
        } else {
          await fs.copy(fullPath, storagePath);
        }
      } else {
        await fs.copy(fullPath, storagePath);
      }

      await storageManager.addVersion(relPath, {
        versionId: uuidv4(),
        timestamp,
        size: stat.size,
        status: 'synced',
        storagePath,
        encoding
      });
      console.log(`[Baseline] Created for: ${relPath}`);
    } catch (e) {}
  }

  async restoreVersion(fileId, versionId, opts = {}) {
    const { asCopy = false, targetPath } = opts;
    const file = storageManager.getFile(fileId);
    if (!file) throw new Error('File not found');

    const version = file.versions.find(v => v.versionId === versionId);
    if (!version) throw new Error('Version not found');
    if (!version.storagePath) throw new Error('This version has no stored file (deletion marker)');

    const config = storageManager.getConfig();
    const watchPath = config.watchPath;
    const originalFull = path.join(watchPath, file.relativePath);

    let destination;
    if (targetPath) {
      destination = targetPath;
    } else if (asCopy) {
      const ext = path.extname(file.name);
      const base = path.basename(file.name, ext);
      const dir = path.dirname(originalFull);
      destination = path.join(dir, `${base}_restored_${Date.now()}${ext}`);
    } else {
      destination = originalFull;
    }

    await fs.ensureDir(path.dirname(destination));
    if (!asCopy && !targetPath) {
      this.pendingRestores.set(file.relativePath, versionId);
    }
    
    if (version.encoding === 'delta') {
      const buffer = await this._reconstructVersion(file, versionId);
      await fs.writeFile(destination, buffer);
    } else if (version.encoding === 'gzip') {
      const gzipped = await fs.readFile(version.storagePath);
      await fs.writeFile(destination, zlib.gunzipSync(gzipped));
    } else {
      await fs.copy(version.storagePath, destination);
    }

    // If restoring to original location, update status
    if (!asCopy && !targetPath) {
      await storageManager.upsertFile(file.relativePath, { 
        currentStatus: 'active',
        lastRestoredVersionId: versionId 
      });
    }

    return { restoredTo: destination };
  }
}

const versionEngine = new VersionEngine();
module.exports = { versionEngine };

File: frontend/src-tauri/src/version_engine.rs
use std::fs;
use std::path::{Path, PathBuf};
use flate2::write::GzEncoder;
use flate2::read::GzDecoder;
use flate2::Compression;
use std::io::{Read, Write};
use crate::storage_manager::{StorageManager, FileRecord, Version};
use uuid::Uuid;
use chrono::Utc;
use fossil_delta;

pub struct VersionEngine;

impl VersionEngine {
    pub fn new() -> Self {
        VersionEngine {}
    }

    fn should_skip_gzip(ext: &str) -> bool {
        let skip = vec![
            "zip", "rar", "7z", "tar", "gz",
            "mp4", "mkv", "avi", "mov",
            "mp3", "flac", "aac",
            "jpg", "jpeg", "png", "gif", "webp"
        ];
        skip.contains(&ext)
    }

    pub fn process_event(
        sm: &mut StorageManager,
        action: &str,
        rel_path: &str,
    ) -> Result<(), String> {
        let wp = match &sm.config.watch_path {
            Some(p) => p.clone(),
            None => return Ok(()),
        };

        let full_path = Path::new(&wp).join(rel_path);
        let storage_dir = Path::new(&wp).join(".restorex").join("versions").join(Path::new(rel_path).parent().unwrap_or(Path::new("")));
        
        if action == "deleted" {
            let meta = sm.ensure_metadata(&wp);
            let timestamp = Utc::now().to_rfc3339();
            if let Some(file_record) = meta.files.get_mut(rel_path) {
                file_record.current_status = "deleted".to_string();
                file_record.last_seen = timestamp.clone();
                file_record.versions.push(Version {
                    version_id: Uuid::new_v4().to_string(),
                    timestamp: timestamp.clone(),
                    size: 0,
                    status: "deleted".to_string(),
                    storage_path: None,
                    restored_from: None,
                    encoding: None,
                    parent_version_id: None,
                });
                sm.save_meta();
            }
            return Ok(());
        }
        
        if action == "created" || action == "modified" {
            let metadata = match fs::metadata(&full_path) {
                Ok(m) => m,
                Err(_) => return Ok(()), // file might be deleted quickly
            };

            if !metadata.is_file() {
                return Ok(());
            }

            let _ = fs::create_dir_all(&storage_dir);
            let size = metadata.len();
            let timestamp = Utc::now().to_rfc3339();
            let safe_ts = timestamp.replace(":", "-").replace(".", "-");
            let file_name = Path::new(rel_path).file_name().unwrap().to_str().unwrap();
            let storage_name = format!("__{}__{}", safe_ts, file_name);
            let storage_path = storage_dir.join(storage_name);
            let version_id = Uuid::new_v4().to_string();

            // 1. Read current data
            let mut current_buffer = Vec::new();
            if let Ok(mut file) = fs::File::open(&full_path) {
                let _ = file.read_to_end(&mut current_buffer);
            } else {
                return Ok(());
            }

            // 2. Try to find the latest version to create a delta
            let mut encoding = "full".to_string();
            let mut parent_version_id = None;
            let mut storage_data = current_buffer.clone();

            let mut meta = sm.ensure_metadata(&wp);
            if let Some(file_record) = meta.files.get(rel_path) {
                if let Some(latest_v) = file_record.versions.last() {
                    // Try to reconstruct previous version to create a delta
                    if let Ok(prev_buffer) = Self::reconstruct_buffer(file_record, &latest_v.version_id) {
                        let delta_bytes = fossil_delta::delta(&prev_buffer, &current_buffer);
                        
                        // Check if delta is actually smaller (including GZIP consideration)
                        if delta_bytes.len() < current_buffer.len() / 2 {
                            storage_data = delta_bytes;
                            encoding = "delta".to_string();
                            parent_version_id = Some(latest_v.version_id.clone());
                        }
                    }
                }
            }

            // 3. Fallback to GZIP if still "full" and eligible
            let ext = Path::new(rel_path).extension().unwrap_or_default().to_str().unwrap_or("").to_lowercase();
            if encoding == "full" && size <= 200 * 1024 * 1024 && !Self::should_skip_gzip(&ext) {
                let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
                if encoder.write_all(&current_buffer).is_ok() {
                    if let Ok(gzipped) = encoder.finish() {
                        if gzipped.len() < current_buffer.len() {
                            storage_data = gzipped;
                            encoding = "gzip".to_string();
                        }
                    }
                }
            }

            // 4. Save to disk
            let _ = fs::write(&storage_path, &storage_data);

            // Update metadata
            let meta = sm.ensure_metadata(&wp);
            let file_record = meta.files.entry(rel_path.to_string()).or_insert(FileRecord {
                id: Uuid::new_v4().to_string(),
                name: file_name.to_string(),
                relative_path: rel_path.to_string(),
                current_status: "active".to_string(),
                size: Some(size),
                versions: vec![],
                last_seen: timestamp.clone(),
                last_restored_version_id: None,
            });

            file_record.versions.push(Version {
                version_id,
                timestamp: timestamp.clone(),
                size,
                status: action.to_string(),
                storage_path: Some(storage_path.to_str().unwrap().to_string()),
                restored_from: None,
                encoding: Some(encoding),
                parent_version_id,
            });

            file_record.current_status = action.to_string();
            file_record.last_seen = timestamp;
            file_record.size = Some(size);
            
            sm.save_meta();
        }

        Ok(())
    }

    pub fn restore_version(
        sm: &mut StorageManager,
        file_id: &str,
        version_id: &str,
        as_copy: bool,
    ) -> Result<String, String> {
        let wp = sm.config.watch_path.clone().ok_or("No watch path configured")?;
        let meta = sm.ensure_metadata(&wp);
        
        let file_record = meta.files.values().find(|f| f.id == file_id)
            .ok_or_else(|| format!("File not found in metadata: {}", file_id))?.clone();
            
        let mut current_vid = version_id.to_string();
        let mut patch_stack = Vec::new();
        
        // 1. Follow the delta chain upwards to the baseline
        loop {
            let v = file_record.versions.iter().find(|v| v.version_id == current_vid)
                .ok_or_else(|| format!("Version {} not found in timeline", current_vid))?;
            
            patch_stack.push(v.clone());
            
            if let Some(parent_id) = &v.parent_version_id {
                current_vid = parent_id.clone();
            } else {
                break; // Found the baseline (full file or gzip)
            }
        }
        
        // 2. Load and process baseline (the last item in stack)
        let baseline_v = patch_stack.pop().unwrap();
        let mut current_buffer = Vec::new();
        
        if let Some(sp) = baseline_v.storage_path {
            let bytes = fs::read(&sp).map_err(|e| format!("Failed to read baseline: {}", e))?;
            if baseline_v.encoding.as_deref() == Some("gzip") {
                let mut decoder = GzDecoder::new(&bytes[..]);
                let mut decoded = Vec::new();
                decoder.read_to_end(&mut decoded).map_err(|e| format!("Gzip decode failed: {}", e))?;
                current_buffer = decoded;
            } else {
                current_buffer = bytes;
            }
        } else {
            return Err("Baseline version has no storage path".into());
        }
        
        // 3. Apply patches in reverse order (bottom-up from baseline)
        patch_stack.reverse();
        for p in patch_stack {
            if let Some(sp) = p.storage_path {
                let delta_bytes = fs::read(&sp).map_err(|e| format!("Failed to read delta: {}", e))?;
                current_buffer = fossil_delta::apply(&current_buffer, &delta_bytes).map_err(|e| format!("Delta apply failed: {:?}", e))?;
            }
        }
        
        // 4. Determine destination path
        let dest_path = if as_copy {
            let original_name = &file_record.name;
            let stem = Path::new(original_name).file_stem().unwrap_or_default().to_str().unwrap_or("restored");
            let ext = Path::new(original_name).extension().unwrap_or_default().to_str().unwrap_or("");
            let ts = Utc::now().timestamp();
            Path::new(&wp).join(format!("{}_restored_{}.{}", stem, ts, ext))
        } else {
            Path::new(&wp).join(&file_record.relative_path)
        };
        
        // 5. Write the reconstructed file
        if let Some(parent) = dest_path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        fs::write(&dest_path, current_buffer).map_err(|e| format!("Failed to write restored file: {}", e))?;
        
        // 6. Update metadata with a "restored" event if not a copy
        if !as_copy {
            let timestamp = Utc::now().to_rfc3339();
            let wp = sm.config.watch_path.clone().unwrap_or_default();
            let meta = sm.ensure_metadata(&wp);
            if let Some(file_record) = meta.files.get_mut(&file_record.relative_path) {
                file_record.current_status = "restored".to_string();
                file_record.last_seen = timestamp.clone();
                file_record.last_restored_version_id = Some(version_id.to_string());
                file_record.versions.push(Version {
                    version_id: Uuid::new_v4().to_string(),
                    timestamp: timestamp.clone(),
                    size: file_record.size.unwrap_or(0),
                    status: "restored".to_string(),
                    storage_path: None,
                    restored_from: Some(version_id.to_string()),
                    encoding: None,
                    parent_version_id: None,
                });
                sm.save_meta();
            }
        }

        Ok(format!("Successfully restored to {}", dest_path.display()))
    }

    pub fn restore_folder(
        sm: &mut StorageManager,
        folder_path: &str,
    ) -> Result<String, String> {
        let wp = sm.config.watch_path.clone().ok_or("No watch path")?;
        let meta = sm.ensure_metadata(&wp).clone(); // Clone to avoid borrow issues in loop
        
        let files_to_restore: Vec<String> = meta.files.values()
            .filter(|f| f.relative_path.starts_with(folder_path) && f.current_status == "deleted")
            .map(|f| f.id.clone())
            .collect();

        if files_to_restore.is_empty() {
            return Ok("No deleted files found in this folder".into());
        }

        let mut count = 0;
        for fid in files_to_restore {
            let file = meta.files.values().find(|f| f.id == fid).unwrap();
            // Find latest version with storage
            if let Some(v) = file.versions.iter().rev().find(|v| v.storage_path.is_some()) {
                if let Ok(_) = Self::restore_version(sm, &fid, &v.version_id, false) {
                    count += 1;
                }
            }
        }

        Ok(format!("Successfully restored {} files in {}", count, folder_path))
    }

    // Helper to reconstruct buffer for internal use
    fn reconstruct_buffer(file_record: &FileRecord, version_id: &str) -> Result<Vec<u8>, String> {
        let mut current_vid = version_id.to_string();
        let mut patch_stack = Vec::new();
        
        loop {
            let v = file_record.versions.iter().find(|v| v.version_id == current_vid)
                .ok_or_else(|| format!("Version {} not found", current_vid))?;
            
            patch_stack.push(v.clone());
            
            if let Some(parent_id) = &v.parent_version_id {
                current_vid = parent_id.clone();
            } else {
                break;
            }
        }
        
        let baseline_v = patch_stack.pop().unwrap();
        let mut current_buffer = Vec::new();
        
        if let Some(sp) = baseline_v.storage_path {
            let bytes = fs::read(&sp).map_err(|e| e.to_string())?;
            if baseline_v.encoding.as_deref() == Some("gzip") {
                let mut decoder = GzDecoder::new(&bytes[..]);
                let _ = decoder.read_to_end(&mut current_buffer);
            } else {
                current_buffer = bytes;
            }
        } else {
            return Err("No baseline".into());
        }
        
        patch_stack.reverse();
        for p in patch_stack {
            if let Some(sp) = p.storage_path {
                let delta_bytes = fs::read(&sp).map_err(|e| e.to_string())?;
                current_buffer = fossil_delta::apply(&current_buffer, &delta_bytes).map_err(|e| format!("{:?}", e))?;
            }
        }
        
        Ok(current_buffer)
    }
}
